src/plots/plots.js

```diff
@@ -609,7 +592,61 @@ function relinkPrivateKeys(toLayout, fromLayout) {
     }
 }
 
-plots.supplyDataDefaults = function(traceIn, i, layout) {
+plots.supplyDataDefaults = function(dataIn, dataOut, layout, modules) {
+    var cnt = 0;
+
+    // push to array if item isn't already in array
+    function fill(arr, item) {
+        if(item && arr.indexOf(item) === -1) return arr.push(item);
+    }
+
+    // detect plot type
+    function detect(trace) {
+        if(plots.traceIs(trace, 'cartesian')) layout._hasCartesian = true;
+        else if(plots.traceIs(trace, 'gl3d')) layout._hasGL3D = true;
+        else if(plots.traceIs(trace, 'geo')) layout._hasGeo = true;
+        else if(plots.traceIs(trace, 'pie')) layout._hasPie = true;
+        else if(plots.traceIs(trace, 'gl2d')) layout._hasGL2D = true;
+        else if(plots.traceIs(trace, 'ternary')) layout._hasTernary = true;
+        else if('r' in trace) layout._hasPolar = true;
+    }
+
+    for(var i = 0; i < dataIn.length; i++) {
+        var trace = dataIn[i];
+
+        var fullTrace = plots.supplyTraceDefaults(trace, cnt, layout);
+
+        // keep track of pre-transform _input
+        var traceIn = fullTrace._input;
+
+        if(fullTrace.transforms && fullTrace.transforms.length) {
+            var expandedTraces = applyTransforms(fullTrace, dataOut, layout);
+
+            for(var j = 0; j < expandedTraces.length; j++) {
+                var expandedTrace = expandedTraces[j];
+                var fullExpandedTrace = plots.supplyTraceDefaults(expandedTrace, cnt, layout);
+
+                // copy refs
+                fullExpandedTrace._input = traceIn;
+                fullExpandedTrace._fullTransforms = fullTrace.transforms;
+                fullExpandedTrace._index = i;
```

Comment:
not sure what we need `fullTrace.index` for - as provided below in `supplyTraceDefaults` it's `cnt`, but we should give these better names than `index` and `_index` - maybe `_inputIndex` and `_expandedIndex`? I can imagine needing both in different cases...